### DISCUSSION TITLE
Abstract Const & Const Generics

### DISCUSSION THREAD

**Discussion by YT-GameMaker** [dotnet/csharplang#7648] [Language Ideas]
I just saw this: [#7508](https://github.com/dotnet/csharplang/discussions/7508), it is a great proposal, but before reading it, I thought of a simple syntactic sugar, no need to modify the existing type system. The example code is roughly as follows.

```
    // Example ////////////

    class UserType<T1, int x, T2, int y> { }
    void Func() => new UserType<string, 5, float, 16>();


    // Translated ////////////

    // This is a built-in interface
    interface IConstIntGeneric { abstract const int Value; }

    class UserType<T1, TGen1, T2, TGen2>
        where TGen1 : IConstIntGeneric
        where TGen2 : IConstIntGeneric
    {
        const int x = TGen1.Value;
        const int y = TGen2.Value;
    }

    struct Gen_int_5 : IConstIntGeneric { public const int Value = 5; }
    struct Gen_int_16 : IConstIntGeneric { public const int Value = 16; }

    void Func() => new UserType<string, Gen_int_5, float, Gen_int_16>();
```
As you can see, it almost exclusively uses the existing C#, except for `abstract const`. When `x` and `y` are applied to the generic parameter, they are replaced with the generated struct; otherwise, they are treated as constants inside the user type.

This idea was inspired by `abstract static`, which is mainly used in conjunction with generics, and the same goes for `abstract const`. But constants are determined at compile time, while static variables are not. We know that constructed generics are also determined at compile time, and `abstract const` has linked generics and constants together. So, by this syntactic sugar, we can merge them into the same thing. If constants can be used as types, they can be passed as generic parameters like types, thus achieving code sharing. For example, we can create something like `InlineArray<T, int size>`.

**Comment by HaloFour** [comment]
Both would require considerable runtime changes, given `abstract const` is not a thing.

**Comment by CyrusNajmabadi** [comment]
Closing as a dupe of the linked issue, which already covers these concepts

**Reply by YT-GameMaker** [reply]
My idea is `abstract const` + syntactic sugar, by adding interfaces to avoid introducing new type about const generics (which would affect reflection), it's a different solution, please reopen it!

**Reply by CyrusNajmabadi** [reply]
Yes. As stated, it's effectively a duplicate of the other topic, just with a tiny syntactic tweak. We don't need a separate discussion for that. It can happen in the main discussion thread. Thanks.
